homeassistant/components/sonarr/coordinator.py

```diff
@@ -0,0 +1,148 @@
+"""Data update coordinator for the Sonarr integration."""
+from __future__ import annotations
+
+from abc import abstractmethod
+from datetime import timedelta
+from typing import TypeVar, Union, cast
+
+from aiopyarr import (
+    Command,
+    Diskspace,
+    SonarrCalendar,
+    SonarrQueue,
+    SonarrSeries,
+    SonarrWantedMissing,
+    exceptions,
+)
+from aiopyarr.models.host_configuration import PyArrHostConfiguration
+from aiopyarr.sonarr_client import SonarrClient
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+import homeassistant.util.dt as dt_util
+
+from .const import CONF_UPCOMING_DAYS, CONF_WANTED_MAX_ITEMS, DOMAIN, LOGGER
+
+T = TypeVar(
+    "T",
+    bound=Union[
+        list[SonarrCalendar],
+        list[Command],
+        list[Diskspace],
+        SonarrQueue,
+        list[SonarrSeries],
+        str,
+        SonarrWantedMissing,
+    ],
+)
+
+
+class SonarrDataUpdateCoordinator(DataUpdateCoordinator[T]):
+    """Data update coordinator for the Sonarr integration."""
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        host_configuration: PyArrHostConfiguration,
+        api_client: SonarrClient,
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass=hass,
+            logger=LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=30),
+        )
+        self.api_client = api_client
+        self.host_configuration = host_configuration
+        self.system_version: str | None = None
+
+    async def _async_update_data(self) -> T:
+        """Get the latest data from Sonarr."""
+        try:
+            return await self._fetch_data()
+
+        except exceptions.ArrConnectionException as ex:
+            raise UpdateFailed(ex) from ex
+        except exceptions.ArrAuthenticationException as ex:
+            raise ConfigEntryAuthFailed(
+                "API Key is no longer valid. Please reauthenticate"
+            ) from ex
+
+    @abstractmethod
+    async def _fetch_data(self) -> T:
+        """Fetch the actual data."""
+        raise NotImplementedError
```

Comment:
Either use `@abstractmethod` (but then also inherit from `abc.ABC`), or do `raise NotImplementedError` (without `abstractmethod`). The later is the one used by core integrations.

https://github.com/home-assistant/core/blob/a18a0b39dd2e6599741633c27ee8c43b488cc460/homeassistant/helpers/entity.py#L978-L981